cabal-install/bootstrap.py

```diff
@@ -0,0 +1,299 @@
+        subprocess.run([str(ghc_pkg_path), 'describe', f'{dep.package}-{dep.version}'],
+                       check=True, stdout=subprocess.DEVNULL)
+        print(f'Using {dep.package}-{dep.version} from GHC...')
+        return
+
+    elif dep.source == PackageSource.HACKAGE:
+        tarball = fetch_package(dep.package, dep.version, dep.revision, dep.sha256)
+        subprocess.run(['tar', 'xf', tarball.resolve()],
+                       cwd=PACKAGES, check=True)
+        sdist_dir = PACKAGES / f'{dep.package}-{dep.version}'
+
+        # Update cabal file with revision
+        if dep.revision is not None:
+            shutil.copyfile(PACKAGES / f'{dep.package}.cabal',
+                            sdist_dir / f'{dep.package}.cabal')
+
+    elif dep.source == PackageSource.LOCAL:
+        if dep.package == 'Cabal':
+            sdist_dir = Path('Cabal').resolve()
+        elif dep.package == 'cabal-install':
+            sdist_dir = Path('cabal-install').resolve()
+        else:
+            raise 'hi'
+
+    install_sdist(sdist_dir)
+
+def install_sdist(sdist_dir: Path):
+    prefix = (PACKAGES / 'tmp').resolve()
+    configure_args = [
+        f'--package-db={PKG_DB.resolve()}',
+        f'--prefix={prefix}',
+        f'--with-compiler={ghc_path}',
+        f'--with-hc-pkg={ghc_pkg_path}',
+    ]
+
+    def check_call(args: List[str]) -> None:
+        subprocess.run(args, cwd=sdist_dir, check=True)
+
+    check_call([str(ghc_path), '--make', 'Setup'])
+    check_call(['./Setup', 'configure'] + configure_args)
+    check_call(['./Setup', 'build'])
+    check_call(['./Setup', 'install'])
+
+def hash_file(h, f: BinaryIO) -> bytes:
+    while True:
+        d = f.read(1024)
+        if len(d) == 0:
+            return h.hexdigest()
+
+        h.update(d)
+
+
+# Cabal plan.json representation
+UnitId = NewType('UnitId', str)
+PlanUnit = NewType('PlanUnit', object)
+
+def read_plan(project_dir: Path) -> Dict[UnitId, PlanUnit]:
+    path = project_dir / 'dist-newstyle' / 'cache' / 'plan.json'
+    plan = json.load(path.open('rb'))
+    return {
+        UnitId(c['id']): PlanUnit(c)
+        for c in plan['install-plan']
+    }
+
+def extract_plan() -> None:
+    units = read_plan(Path('.'))
+    target_unit = [
+        unit
+        for unit in units.values()
+        if unit['pkg-name'] == 'cabal-install'
+        if unit['component-name'] == 'exe:cabal'
+    ][0]
+
+    def unit_to_bootstrap_dep(unit: PlanUnit) -> BootstrapDep:
+        if 'pkg-src' in unit and unit['pkg-src']['type'] == 'local':
+            source = PackageSource.LOCAL
+        elif unit['type'] == 'configured':
+            source = PackageSource.HACKAGE
+        elif unit['type'] == 'pre-existing':
+            source = PackageSource.PREEXISTING
+
+        return BootstrapDep(package = unit['pkg-name'],
+                            version = unit['pkg-version'],
+                            source = source,
+                            revision = None,
```

Comment:
This is TODO?

Reply:
Indeed. This is the largest issue at the moment. I don't believe this information is accessible from `plan.json.

Reply:
@phadej, how best to go about addressing this?

Reply:
The only approach I know (which doesn't require changing `cabal`) is to scan through `01-index.tar` and find the info there. That's how `cabal-bundler` works.

The `cabal` doesn't really know which revision it's working with, as the number is not defining one. We probably know the sha256 of the revision, and I'd rather change hackage-server to provide revisions based on the hash. The revision number is for humans, not the machines.

Reply:
Alright, let me know how this looks.